### PR TITLE
Postprocessing (WIP)

### PR DESCRIPTION
**Pull request by RealBadAngel**
Goal: add shaders postprocessing effects

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -1795,6 +1803,16 @@ void Game::run()
...
 
+	std::string pp_shaders_path = std::string("client") + DIR_DELIM
+				+ "shaders" + DIR_DELIM
+				+ "postprocess" + DIR_DELIM;
+
+	irrPP = createIrrPP(device, video::EPQ_FULL, pp_shaders_path.c_str());
+	mrt = new Mrt(device);
+
+	irr::video::CPostProcessingEffectChain* pp = irrPP->createEffectChain();
+    pp->createEffect(video::EPE_DOF)->addTextureToShader(mrt->getDepthRTT());
```
Comment (nerzhul) [inline src/game.cpp]: indent

File: src/game.cpp [hunk 2]
```diff
@@ -4093,7 +4111,7 @@ void Game::updateFrame(std::vector<aabb3f> &highlight_boxes,
 		stats->beginscenetime = timer.stop(true);
 	}
 
-	draw_scene(driver, smgr, *camera, *client, player, *hud, *mapper,
+	draw_scene(driver, smgr, *camera, *client, player, *hud, *mapper, *mrt, *irrPP,
```
Comment (nerzhul) [inline src/game.cpp]: why not pass pointer for mrt instead of reference ?

File: src/irrpp/IShaderDefaultPostProcessCallback.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		irr::video::IMaterialRendererServices* services,
+		irr::s32 userData)
+{
+    irr::s32 render = 0;
+	irr::s32 tex0 = 1;
+    irr::s32 tex1 = 2;
+    irr::s32 tex2 = 3;
+
+    services->setPixelShaderConstant("Render", &render, 1);
+	services->setPixelShaderConstant("Tex0", &tex0, 1);
```
Comment (nerzhul) [inline src/irrpp/IShaderDefaultPostProcessCallback.cpp]: indent

**Comment by MirceaKitsune** [comment]
Just a heads up: As most of you probably know by now, RealBadAngel died yesterday due to complications after his heart attack two weeks ago. RIP, you did some amazing things for Minetest!

Back on the topic of the pull request: If someone wishes the code to be finished, we will need new developers to work on it. I spoke frequently with RealBadAngel in PM about his new shaders, but never looked at the actual code and can't say I know it. Both for the project but also in his memory, I'm hoping that someone else will finish what he started and get this merged into core, once the feature is in an acceptable form.

**Comment by paramat** [comment]
Would anyone like to take this over in a new PR so we can close this?

**Comment by tobyplowy** [comment]
@paramat I think you can close this ( it doesn't seem anyone is picking it up😞)

**Comment by paramat** [comment]
I'd rather not, it's a major piece of work that needs to be visible to encourage someone to take over.

**Comment by paramat** [comment]
Perhaps we can give this a year and close it in Feb 2017? Or is it so valuable it should wait for a new shader dev to come along?

**Comment by nerzhul** [comment]
I also think we should close this as nobody works on it an a massive conflict is present on this PR

**Comment by paramat** [comment]
Yes i agree to a close, the commits will still be findable if someone wants to take over.

**Comment by lhofhansl** [comment]
I'll look at some of these and see if/when/how I can finish some of them.

**Comment by paramat** [comment]
@lhofhansl i somewhat regret asking you to consider this.
Effects like this depth of field are in my opinion overly fancy for minetest, maybe bloom too, although i like bloom in other contexts (like anime).
I think there are much more important things to work on and i would recommend directing your efforts elsewhere.
A problem with adopting this work is that it is very tied-in to RBAs former plans for shaders and effects, now of course those plans are gone, much related work will not go ahead, everything has to be questioned and reconsidered.
Another point is that RBAs code is sometimes low quality and problematic.
I feel that if post processing work is continued it is best to re-work it completely. However, being short on dev time, we should work on other things.
For those reasons i'm not sure this should have the 'adoption needed' label.

**Comment by tobyplowy** [comment]
@paramat I think depth of field is really awesome for certain texture packs like summerfields for example but that's just my opinion

**Comment by tobyplowy** [comment]
@paramat also a lot of Minecraft players use shader (mods a lot of them) and if this really is an engine and Not a game it should support it

**Comment by sofar** [comment]
Actually taking DoF out of this PR and isolating it and adding it separately (without the nodes shader stuff) would have my support. DoF + tonemapping is really quite appealing visually.

The node shader stuff (needed for water reflections I think) probably should get more attention to actually be useful.

The box outline shader can probably be dropped entirely.

**Comment by paramat** [comment]
Eww water reflections .. that's certainly far too heavyweight and fancy.

**Comment by nerzhul** [comment]
@paramat i don't think, terasology has beautiful reflections

**Comment by MirceaKitsune** [comment]
Sadly, a lot of people in the Minetest community don't understand the importance of shaders and improved visual effects. I worry that because of the general mentality and "culture" here, the project might forever remain limited to the visuals of stock Minecraft.

**Comment by sofar** [comment]
>Eww water reflections .. that's certainly far too heavyweight and fancy.

For you, maybe. But they are technically not very complex, not should there be any reason to stop Minetest-mods from having (default disabled) water reflections.

Having a small amount of progress ongoing in Minetest's engine visuals is crucial to making sure that we keep developers around that can not only try new fancy things, but also fix existing bugs.

If you say no to all things "too fancy", soon you'll have no developers left who can even fix the most basic bugs in the render pipeline.

So, bring it on.

**Comment by Fixer-007** [comment]
Shaders itself are cool, it is just slow on my ATI HD6870. I was testing MTR_5 branch by RBA with him before his death, and it was close to 30fps, with DOF and other fancy stuff...

Even with basic shader some people having big fps drop offs.

**Comment by paramat** [comment]
Depth of field has a problem in games. It's lovely in anime, but there it is used to direct the viewer's attention to a part of a scene.
In a game we can't predict what distance a player wants to be focussed on, and it causes eyestrain if they look at a distance not in focus. Players need to be able to instantly look at any part of the screen and have it in focus and clear.

How would we decide what range is in focus? A player may have view distance 256 and be looking at something 4 nodes away, or, may have view range 64 and be looking at the furthest visible terrain. So it can't be linked to view range.

This PR, as far as i know, decides what is mid-range from what the crosshair is pointing at, but we do not play a game with our eyes fixed on the very centre of the screen. We could be watching another player travelling in the distance and our view may cross the trunk of a nearby tree, we would not want the focus to shift to the trunk for a moment.
The last time i discussed this PR with RBA the depth of field was badly implemented, it did not get larger at larger distance.

**Comment by Fixer-007** [comment]
I was having 30fps @ 240 v_range in SP with any of his shaders in MTR_5 branch, it is not just DOF one. Without PR4530 or successors it is even worse.

**Comment by sofar** [comment]
@Fixer-007 I wouldn't accept a shader if it meant performance WITH SHADERS DISABLED would be worse.

Also, you are one of the few people that plays/tests on really old hardware. That may not adequately represent the average hardware, so while I think it's great to figure out how to do better, I don't think that depleting minetest visual features is a solution: that kind of canabalization will destroy the project in the long term.

@paramat I didn't say it was perfect, which is why it's best to isolate it to a single PR that can be judged by its merits. This PR has too many things combined already. Splitting it up may just present DoF as a really trivial change that is worth it to someone (we're had 2 new folks do some nice things recently). So let's not discourage them ?

Let the code speak before dismissing things with subjective negative comments, please.

**Comment by Fixer-007** [comment]
>I wouldn't accept a shader if it meant performance WITH SHADERS DISABLED would be worse.

no, I mean performance was even more worse than with usual shaders. My hardware is not that old, it is not 90-ties or 20-ties were PC progress was so fast and difference in 1yr was quite big between generations, right now it is much much more slower, I can spend 600usd again and I will end up with +20% faster CPU and +50% faster GPU at most. It maybe 2011 PC, but it is still pretty fast PC with i3-2120 3.3GHz + ATI HD6870 1GB, from what I see people buy crappy notebooks with intel integrated unfortunately.

**Comment by tobyplowy** [comment]
There's a lot of Minecraft players which say Minecraft looks like shit if you can get the graphics in minetest we could appeal to that audience and potentially get more players, also it Looks good :)
+1 for nice effects :)

@paramat Sorry but I think we have to Branch out You're an Amazing guy/girl but you have to understand minetest must move on otherwise It will get erelivant

**Comment by paramat** [comment]
Sure we can consider DoF, but the problem with it i explained above is not subjective. It's inherently unusable.

**Comment by Fixer-007** [comment]
Last time I checked bloom it was insanely bloomy, cataract-like.

**Comment by paramat** [comment]
No problem, RBA tended to turn everything up to 11, it can be made more subtle.

**Comment by HybridDog** [comment]
>There may be other good post-processing effects.

You could apply slight HSV noise and then perform automatic brightness adjustment on the result.

**Comment by rubenwardy** [comment]
I personally really like water shaders and reflections as I think they're beautiful. I'm not as much as a voxel purist as @paramat though - the furthest I go it thinking that nodes/nodeboxes/node meshes should be cubiod/voxelly. I think that the most important thing for graphics is to maintain immersion, and for me water reflections do that.

I love this video: https://www.youtube.com/watch?v=Rkydse1LVX0

I think that DOF is done very well in Skyrim, and isn't distracting at all. However, if it was to be added to MT it would need to be optional. Just like any other shader effect.

**Comment by MirceaKitsune** [comment]
@rubenwardy Wonderful video there, thanks for sharing that as an example! That is indeed the way I would like to see a modern voxel engine evolve, and how I was dreaming to see Minetest itself become at some point. Unless other people in the team can be determined to see the importance of immersive visuals like that, and understand that you can't have a true experience by staying limited to 16bit era visuals... this might sadly not happen.

**Comment by tobyplowy** [comment]
@rubenwardy that video Looks So Beautiful :') If Only minetest culd do this

**Comment by tobyplowy** [comment]
@rubenwardy  that video Looks So Beautiful :') If Only minetest chuld do this

**Comment by rubenwardy** [comment]
Rebased version here: https://github.com/rubenwardy/minetest/tree/postprocessing

Not offering to maintain it

**Comment by paramat** [comment]
I think the cell shader is unnecessary.
The depth of field maths need improving, currently they don't make the depth of field deeper with distance, so it makes things look miniature through the 'tilt shift' effect:

crossroads

**Comment by MirceaKitsune** [comment]
Someone is working on this again? Thank you! Please post more screenshots as there's progress... I would so like to see more.

**Comment by rubenwardy** [comment]
Yeah, I agree that the cell shading is not really needed.
The DOF is rather distracting in its current form and needs improving

**Comment by sofar** [comment]
- the bloom us unusually bright (I lose all detail)
- the dof is too large, it should be toned down significantly (but I still like it)
- dof and opaque tree leaves are a problem (may be unfixable)
- mrt is probably going to have to stay
- not sure if we want to keep node shaders, but, why not? it opens up a bunch of interesting things.
- cell shader should be removed, I think

that's what I remember is the things to look at.

**Comment by tobyplowy** [comment]
I personally really love cell shaders

**Comment by MirceaKitsune** [comment]
The cell shader is one of the effects I care for less, unlike bloom or depth-of-field which I consider an absolute must. Still I see no reason why it couldn't be accepted, if it can be individually toggled... some people like it, and the cartoon effect does match the blocky style of Minetest pretty well.

**Comment by tobyplowy** [comment]
@MirceaKitsune totally agree it's not super important but I Would love to have it, it wuld look amazing with Summerfields!

**Comment by rubenwardy** [comment]
this is out of date and superceded by #5446 and also my rebase

**Pull request closed without merging**